portal/server/controllers/utilisation-report-service/utilisation-report-upload/utilisation-report-cell-validators/generate-total-fees-accrued-exchange-rate-error.test.js

```diff
@@ -0,0 +1,198 @@
+        value: 'USD',
+        column: 2,
+        row: 1,
+      },
+      'accrual exchange rate': {
+        value: '0.734',
+        row: 1,
+        column: 3,
+      },
+    };
+
+    const exchangeRateError = generateTotalFeesAccruedExchangeRateError(csvDataRow);
+
+    expect(exchangeRateError).toEqual(null);
+  });
+
+  it('returns an error if accrual currency is different to base currency and exchange rate is null', async () => {
+    const csvDataRow = {
+      exporter: {
+        value: testExporterName,
+        column: 1,
+        row: 1,
+      },
+      'accrual currency': {
+        value: 'GBP',
+        column: 2,
+        row: 1,
+      },
+      'base currency': {
+        value: 'USD',
+        column: 2,
+        row: 1,
+      },
+      'accrual exchange rate': {
+        value: null,
+        row: 1,
+        column: 3,
+      },
+    };
+
+    const expectedError = {
+      errorMessage: 'Accrual exchange rate must have an entry when a accrual currency is supplied',
+      column: 3,
+      row: 1,
+      value: null,
+      exporter: testExporterName,
+    };
+
+    const exchangeRateError = generateTotalFeesAccruedExchangeRateError(csvDataRow);
+
+    expect(exchangeRateError).toEqual(expectedError);
+  });
+
+  it('returns an error if accrual currency is different to base currency and accrual exchange rate is not a number', async () => {
+    const csvDataRow = {
+      exporter: {
+        value: testExporterName,
+        column: 1,
+        row: 1,
+      },
+      'accrual currency': {
+        value: 'GBP',
+        column: 2,
+        row: 1,
+      },
+      'base currency': {
+        value: 'USD',
+        column: 2,
+        row: 1,
+      },
+      'accrual exchange rate': {
+        value: 'abc',
+        row: 1,
+        column: 3,
+      },
+    };
+
+    const expectedError = {
+      errorMessage: 'Accrual exchange rate must be a number',
+      column: 3,
+      row: 1,
+      value: 'abc',
+      exporter: testExporterName,
+    };
+
+    const exchangeRateError = generateTotalFeesAccruedExchangeRateError(csvDataRow);
+
+    expect(exchangeRateError).toEqual(expectedError);
+  });
+
+  it('returns an error if accrual currency is different to base currency and exchange rate is too long', async () => {
+    const csvDataRow = {
+      exporter: {
+        value: testExporterName,
+        column: 1,
+        row: 1,
+      },
+      'accrual currency': {
+        value: 'GBP',
+        column: 2,
+        row: 1,
+      },
+      'base currency': {
+        value: 'USD',
+        column: 2,
+        row: 1,
+      },
+      'accrual exchange rate': {
+        value: '1.738491847362543',
+        row: 1,
+        column: 3,
+      },
+    };
+
+    const expectedError = {
+      errorMessage: `Accrual exchange rate must be ${FILE_UPLOAD.MAX_CELL_CHARACTER_COUNT} characters or less`,
+      column: 3,
+      row: 1,
+      value: '1.738491847362543',
+      exporter: testExporterName,
+    };
```

Comment:
I'm glad this behaviour is included explicitly as I was concerned about having so many digits in the exchange rate. Not sure if this is an issue or not, but what if someone accidentally includes extra whitespace characters? Would this error still be thrown or does it trim any leading and trailing whitespace?

Reply:
When we parse the csv we trim off whitespace so wouldn't be shown for leading/trailing, but if there was whitespace in the middle of the number then yes we would have the error :)